File: php/trait-farm.php
<?php

trait Animal
{
	abstract public function noise ();
	
	public function speak () {
		echo $this->noise(), "\n";
	}
}

trait Vehicle
{
	public function go () {
		echo "Brrrrm!", "\n";
	}
}

trait Hitchable
{
	abstract public function go ();
	
	public function pull_plough () {
		echo "Hitching plough", "\n";
		$this->go();
	}
}

class Tractor
{
	use Vehicle;
	use Hitchable;
}

class Horse
{
	use Animal;
	use Hitchable;
	
	public function noise () {
		return "Neigh!";
	}
	
	public function go () {
		echo "Clip-clop", "\n";
	}
}

$tractor = new Tractor;
method_exists($tractor, 'pull_plough') && $tractor->pull_plough();

$horse = new Horse;
method_exists($horse, 'pull_plough') && $horse->pull_plough();

File: scala/TraitFarm.scala
#!/bin/sh
exec scala "$0" "$@"
!#

trait Animal
{
	def noise: String;
	def speak = println(noise);
}

trait Vehicle
{
	def go = println("Brrrrm!");
}

trait Hitchable
{
	def go;
	def pull_plough =
		go;
		println("Hitching plough");
}

class Tractor
	extends Vehicle
	with Hitchable {}

class Horse
	extends Animal
	with Hitchable
{
	def noise: String = "Neigh!";
	def go = println("Clip-clop");
}

val tractor = new Tractor;
if (tractor.isInstanceOf[Hitchable])
	tractor.pull_plough;

val horse = new Horse;
if (horse.isInstanceOf[Hitchable])
	horse.pull_plough;

